### ISSUE TITLE
Metabase adds time to dates columns resulting in wrong dates

### ISSUE THREAD

**Issue by florianmonfort** [metabase/metabase#10335]
Hi,

We have been trying to figure this problem out ourselves but no success so far. Our database format is UTC, and when we query data, this is the raw data / date:

![image](https://user-images.githubusercontent.com/1446429/61097567-58954680-a48e-11e9-84c8-1beae206585d.png)

As you can see, in our database our records for the second line - we are using the same query here - is showing a booking_date of 2019-07-19. 

But when we do the same query in our Metabase, we get the following data in second line:

![image](https://user-images.githubusercontent.com/1446429/61097653-aca02b00-a48e-11e9-8498-c1178561a284.png)

There are two problems here:

1. The date is changed to 2019-07-18. Pretty bad when you're viewing booking data to get dates wrong... 🤦‍♂️😂
2. For some reason Metabase is adding a "time" in that date column, even though technically there is no time in there.

For your information, we have set the database in Metabase to use UTC, same as our Database configuration. We use MySQL.

**Comment by flamber** [comment]
Hi @florianmonfort 
Which version of Metabase?
So you've set Metabase "Reporting Timezone" to UTC, and your MySQL database is UTC?
What is the timezone browser?
Have you checked the timezone troubleshooting? https://metabase.com/docs/latest/troubleshooting-guide/timezones.html

**Comment by tomorrownull** [comment]
> Hi @florianmonfort
> Which version of Metabase?
> So you've set Metabase "Reporting Timezone" to UTC, and your MySQL database is UTC?
> What is the timezone browser?
> Have you checked the timezone troubleshooting? https://metabase.com/docs/latest/troubleshooting-guide/timezones.html

Hi @flamber 
The version just upgrade to 0.32.9.
![image](https://user-images.githubusercontent.com/189354/61194361-7e6f5500-a6f3-11e9-8138-17b5c6118e0e.png)

The reservation_date now is right, but reservation_time still  wrong.

**Comment by flamber** [comment]
@florianmonfort 
So you also had problems with earlier version of Metabase or just after you upgraded to 0.32.9?
Seems like one of the columns is using timezone `+08:00`, so is the rest of your MySQL (database and table) running `UTC`, but you only have problem with that single column?

**Comment by tomorrownull** [comment]
> @florianmonfort
> So you also had problems with earlier version of Metabase or just after you upgraded to 0.32.9?
> Seems like one of the columns is using timezone `+08:00`, so is the rest of your MySQL (database and table) running `UTC`, but you only have problem with that single column?

@flamber  
Yes,Just upgraded to 0.32.9.
Seems have problem if type of column is time

This is the table scheme and data.
`
+--------------------------+--------------+------+-----+---------+----------------+
| Field                    | Type         | Null | Key | Default | Extra          |
+--------------------------+--------------+------+-----+---------+----------------+
| reservation_date   | date         | YES  | MUL | NULL    |                |
| reservation_time   | time         | YES  | MUL | NULL    |                |
`


`
mysql> select * from books limit 1 \G;
*************************** 1. row ***************************
     
        reservation_date: 2019-07-01
        reservation_time: 19:30:00

`

**Comment by flamber** [comment]
@florianmonfort 
There was a fix (#9461) in 0.32.9 specifically about how `Time` field type was displayed. So before, it used to display time using browser timezone, but now time is displayed according to Reporting Timezone.

**Comment by paulrosenzweig** [comment]
@tomorrownull, I think the behavior you're seeing is expected.

Just to confirm, you have `19:30:00` stored in `reservation_time`. That column doesn't have timezone data, but the values you store are assumed to be in UTC. Your reporting timezone is also set to UTC. Is that right?

If so, I think you should expect to see this in the UI. 
![image](https://user-images.githubusercontent.com/691495/61229209-c55e4a00-a6f5-11e9-8054-5b3110751b49.png)

I think @flamber is right and that this is collateral of the fix in 0.32.9. What time would you expect to see there? Maybe we making too many assumptions about how time columns should be displayed.

**Comment by florianmonfort** [comment]
Hi @paulrosenzweig ,

To fix it so far we have tried to set the reporting timezone to "database default", so not telling it to use UTC, and then in our query we did the following:

![image](https://user-images.githubusercontent.com/1446429/61268306-9b785680-a7cd-11e9-8346-ad437efd382c.png)

Then time data appears correctly. Both Time (time type) column and created_at (timestamp format) appear correctly:
![image](https://user-images.githubusercontent.com/1446429/61268769-18580000-a7cf-11e9-9a17-d211c863188c.png)

If we use the following query:

![image](https://user-images.githubusercontent.com/1446429/61268347-b1861700-a7cd-11e9-9bcf-afac9e7fb889.png)
With reporting timezone set to Database Default, we get the following result:
![image](https://user-images.githubusercontent.com/1446429/61268365-c662aa80-a7cd-11e9-9a65-7799bcc737c1.png)

Which is not what we want.

If we use the following query:

![image](https://user-images.githubusercontent.com/1446429/61268394-da0e1100-a7cd-11e9-8885-db2441c53620.png)
With reporting timezone set to UTC:
![image](https://user-images.githubusercontent.com/1446429/61268405-eabe8700-a7cd-11e9-8458-0b07e9bb417e.png)

Then we get the following result:

![image](https://user-images.githubusercontent.com/1446429/61268679-beefd100-a7ce-11e9-96a5-55551f9eb249.png)
Time is right, but timestamp column created_at is wrong, this is the actual value in the DB:
![image](https://user-images.githubusercontent.com/1446429/61268719-ea72bb80-a7ce-11e9-8586-12f40763ca5a.png)

So... Each solution that we've tried so far creates problems in either time column or created_at column.

**Comment by paulrosenzweig** [comment]
@florianmonfort Are all three columns (reservation date, reservation time, and created at) stored in UTC? Or, are the reservation columns referring to local time? 

It might be helpful to dig into a specific example. If on July 17th at 14:00 (local time) in Shanghai someone creates a reservation for that night at 19:00, what values appear in the db and what values do you expect to appear in the UI?

**Comment by florianmonfort** [comment]
Here is a test I do for you following an example:

I create a booking in our system:

![image](https://user-images.githubusercontent.com/1446429/61359040-2d579080-a8ae-11e9-8a03-9ff583deacab.png)

Our database in Sequel Pro shows this:

![image](https://user-images.githubusercontent.com/1446429/61359127-5710b780-a8ae-11e9-9752-b0864828e7ac.png)
![image](https://user-images.githubusercontent.com/1446429/61359140-5e37c580-a8ae-11e9-84f5-9ea9b354d356.png)
All information is correct, including on created_at timestamp for the time at which I created the booking.

Now here is what our internal tools display:

![image](https://user-images.githubusercontent.com/1446429/61359240-88898300-a8ae-11e9-9718-49378b56ece2.png)
![image](https://user-images.githubusercontent.com/1446429/61359264-8f17fa80-a8ae-11e9-9879-38343aeb19d5.png)

Date, time etc. all correct.

Now here is the same query/data from Sequel Pro done in Metabase - **IF reporting timezone in Metabase is set to UTC**:

![image](https://user-images.githubusercontent.com/1446429/61359436-e0c08500-a8ae-11e9-8d00-4a81a28e0206.png)
So here for some reason the date column "gets" a time that seems to come out of nowhere, and then on top of that:
![image](https://user-images.githubusercontent.com/1446429/61359491-fafa6300-a8ae-11e9-84c5-eb5a5e6fa0ac.png)
created_at is also wrong.

Now here is the same query/data from Sequel pro done in Metabase - **IF reporting timezone is set to database default**:

![image](https://user-images.githubusercontent.com/1446429/61359577-2bda9800-a8af-11e9-83b7-c59247381172.png)
Here the data also again becomes changed, and then:
![image](https://user-images.githubusercontent.com/1446429/61359609-3d23a480-a8af-11e9-9cf9-c1e510c79484.png)
created_at is correct.

So in conclusion, no matter the configuration, we can't get any result that actually gets all the columns right.

Hope this case explanation is more clear 😁

**Comment by paulrosenzweig** [comment]
Thanks for all the details! I dug into this a bit more, and i think there are two issues at play here.

One issue is that MySQL doesn't come loaded with timezone data. I had just been checking things in postgres. When I tried with the same data in mysql I saw the same problem you did. The created_at and reservation_date would shift.

To fix, I ran this against my instance of mysql:
`mysql_tzinfo_to_sql /usr/share/zoneinfo | mysql -u root mysql`

(solution was found [here](https://discourse.metabase.com/t/solved-time-zone-issue-mysql/1293/3))

That made mysql act like postgres and prevents the weird timeshifting on your date column. If this is the issue you're experiencing, I think adding timezone data to mysql and setting the reporting timezone to UTC should fix your problems.

I'm going to see if there's a good way to automatically alert users that a mysql instance is missing this timezone data.

---

The other issue is that we don't adjust times based on reporting timezone the same way we do datetimes and dates.

For example, with a reporting timezone of UTC, we might return these values for a timestamp, date, and time:
```
"2019-07-17T16:17:14.000Z"
"2019-07-01T00:00:00.000Z"
"19:30:00.000Z"
```
But, with a reporting timezone of Hong Kong, we'd see these values:
```
"2019-07-17T16:17:14.000+08:00"
"2019-07-01T00:00:00.000+08:00"
"03:30:00.000+08:00"
```

For the date and timestamp, we adjust only the timezone, but for time we adjust both the timezone and time itself. I'm still trying to understand this behavior, and see if it can be more consistent.

**Comment by florianmonfort** [comment]
@paulrosenzweig thanks for the reply.

For part 1, we are looking into doing this. We're a bit reluctant because we are afraid of what effect this may have on the million APIs we have that rely on this DB for critical software...

For part 2, good luck with that and I hope you guys can let me know soon when this is done.

**Comment by tomorrownull** [comment]
> Are all three columns (reservation date, reservation time, and created at) stored in UTC? Or, are the reservation columns referring to local time?
> 
> It might be helpful to dig into a specific example. If on July 17th at 14:00 (local time) in Shanghai someone creates a reservation for that night at 19:00, what values appear in the db and what values do you expect to appear in the UI?


Thanks @florianmonfort,@flamber 👍🏻,
Executed this script,
The bug fixed
```
mysql_tzinfo_to_sql /usr/share/zoneinfo | mysql -u root mysql
```

**Comment by florianmonfort** [comment]
@tomorrownull @paulrosenzweig @flamber thanks guys for all your help 👍

**Comment by paulrosenzweig** [comment]
Glad that worked! I'll close this issue now that I've split off the other two.
